### PR TITLE
Fix for create cards and better preservation of object identity during deserialization

### PR DESCRIPTION
**Pull request by habdelra**
This PR fixes a bug where we were unable to create cards from operator mode due to an incorrect relativeTo. Also, this PR improves the preservation of object identity during deserialization--specifically it will preserve object identity in the field values of a card by leveraging our live models (for which we have added a public API for in @cardstack/runtime-common.

### CODE REVIEW

File: packages/host/tests/acceptance/operator-mode-test.ts [hunk 1]
```diff
@@ -604,6 +629,47 @@ module('Acceptance | operator mode tests', function (hooks) {
 
       assert.dom('[data-test-search-sheet]').hasClass('closed');
     });
+
+    test<TestContextWithSave>('can create a card from the index stack item', async function (assert) {
```
Comment (habdelra) [inline packages/host/tests/acceptance/operator-mode-test.ts]: this is the test that was missing that could have caught the original issue

File: packages/host/tests/acceptance/operator-mode-test.ts [hunk 2]
```diff
@@ -1210,11 +1286,6 @@ module('Acceptance | operator mode tests', function (hooks) {
       },
     );
     await waitFor('[data-test-save-idle]');
-    assert.strictEqual(
-      getMonacoContent(),
-      JSON.stringify(expected, null, 2),
-      'monaco content has updated',
-    );
```
Comment (habdelra) [inline packages/host/tests/acceptance/operator-mode-test.ts]: this fixes a flaky test by moving these assertions into the onSave callback. we have other tests for asserting that the monaco content is correct.

File: packages/host/app/components/card-catalog/modal.gts [hunk 3]
```diff
@@ -421,11 +425,11 @@ export default class CardCatalogModal extends Component<Signature> {
     // users will open the card catalog modal and insert a new state into the stack.
     let currentState = this.state;
     if (this.state.request.opts?.createNewCard) {
-      newCard = await this.state.request.opts?.createNewCard(ref, undefined, {
+      newCard = await this.state.request.opts?.createNewCard(ref, relativeTo, {
```
Comment (habdelra) [inline packages/host/app/components/card-catalog/modal.gts:428]: this was the critical fix for the bug that prevented card creation from working

File: packages/host/app/components/operator-mode/container.gts [hunk 4]
```diff
@@ -467,8 +478,13 @@ export default class OperatorModeContainer extends Component<Signature> {
...
       },
       viewCard: async (card: CardDef, format: Format = 'isolated') => {
+        let liveCard = await here.cardService.loadModel(here, new URL(card.id));
+        if (!liveCard) {
+          throw new Error(`bug: could not load card ${card.id}`);
+        }
         here.addToStack({
-          card,
+          // assert that we are using a live card
+          card: liveCard,
```
Comment (habdelra) [inline packages/host/app/components/operator-mode/container.gts]: this change ensures that the card we are viewing is the live card--which also exposed the bug that the inner object identities (field values) of the `card-api.updateFromSerialized()` was not preserving object identities all the way down (our existing operator mode integration tests revealed this issue).

File: packages/base/card-api.gts [hunk 5]
```diff
@@ -878,8 +879,11 @@ class LinksTo<CardT extends CardDefConstructor> implements Field<CardT> {
...
     }
-    let cachedInstance = identityContext.identities.get(value.links.self);
+    let cachedInstance =
+      (await getLiveCard(doc, new URL(value.links.self, relativeTo))) ??
+      identityContext.identities.get(value.links.self);
     if (cachedInstance) {
+      cachedInstance[isSavedInstance] = true;
       return cachedInstance as BaseInstanceType<CardT>;
     }
     let resourceId = new URL(value.links.self, relativeTo).href;
```
Comment (habdelra) [inline packages/base/card-api.gts]: this (and the similar linksToMany update) ensures that when possible we re-use the same instance that has already been created for a inner card field value as opposed to always making a new instance.

**Comment by backspace** [comment]
I’ll look at this more thoroughly tomorrow but when I run locally it’s saving two cards:

bm2 2023-10-23 16-32-31

I would definitely not rule out local weirdness though 🙃

**Comment by habdelra** [comment]
> I’ll look at this more thoroughly tomorrow but when I run locally it’s saving two cards:
> 
> bm2 2023-10-23 16-32-31
> I would definitely not rule out local weirdness though 🙃

Ok, I’ll dive into that tomorrow

**Comment by habdelra** [comment]
> I’ll look at this more thoroughly tomorrow but when I run locally it’s saving two cards:
> 
> bm2 2023-10-23 16-32-31
> I would definitely not rule out local weirdness though 🙃

it looks like the issue is that the 2nd save started before the first save completed, and as a result there was not yet an id recieved from the first save. because of that the second save was treated as a create instead of as an update. We'll need to add some additional logic to be able to tell if there is already a save in flight for a card that has been requested to be saved--or probably just an enqueueTask should do the job...

**Comment by habdelra** [comment]
@backspace the issue you found should be fixed now in the latest commit

**Review by backspace — APPROVED** [review]
Confirmed that the double-save doesn’t happen anymore for me locally 🎉

The merge conflict should be easily resolved 🤞🏻

**Pull request merged**